inappmessaging/src/main/java/com/rakuten/tech/mobile/inappmessaging/runtime/runnable/DisplayMessageRunnable.kt

```diff
@@ -61,12 +71,50 @@ internal class DisplayMessageRunnable(
                             .inflate(
                                     R.layout.in_app_message_slide_up,
                                     null) as InAppMessageSlideUpView
-                    slideUpView.populateViewData(message, imageAspectRatio)
+                    slideUpView.populateViewData(message)
                     hostActivity.addContentView(slideUpView, hostActivity.window.attributes)
                 }
                 else -> {
                 }
             }
         }
     }
+
+    private fun downloadImage(view: InAppMessageBaseView, message: Message) {
+        val url = message.getMessagePayload()?.resource?.imageUrl
+        // hide campaign view (cannot use visibility since Glide callback will not work.)
+        // https://github.com/bumptech/glide/issues/618
+        view.alpha = 0f
+        if (url != null) {
+            Glide.with(view).load(url).addListener(object : RequestListener<Drawable> {
+                override fun onLoadFailed(
+                    e: GlideException?,
+                    model: Any?,
+                    target: Target<Drawable>?,
+                    isFirstResource: Boolean
+                ): Boolean {
+                    Timber.tag(TAG).d("Downloading image failed")
+                    // When image can't be downloaded, remove campaign.
+                    DisplayManager.instance().removeMessage(hostActivity)
```

Comment:
So at this point the campaign message is displayed but not visible right? Is it possible that host app calls `closeMessage()` before calling `removeMessage()`? Would that break anything (e.g. next campaign getting closed prematurely)?

Reply:
No, since calling `closeMessage()` will not display the next campaign then the second `removeMessage()` call is ignored.
In `removeMessage()`, if the campaign is not available (already removed in the UI), it will just ignore it and end.